jax/_src/util.py

```diff
@@ -135,10 +135,13 @@ def partition_list(bs: Sequence[bool], l: Sequence[T]) -> tuple[list[T], list[T]
     lists[b].append(x)
   return lists
 
-def merge_lists(bs: Sequence[bool], l0: Sequence[T], l1: Sequence[T]) -> list[T]:
+def merge_lists(bs: Sequence[bool],
+                l0: Sequence[T1],
+                l1: Sequence[T2]
+                ) -> list[T1 | T2]:
```

Comment:
`merge_lists` appears to be called in places where the two lists have different element types.  This allows it to work in those places.